File: backend/models/GeoJSONExport.js
const mongoose = require('mongoose');

const geoJSONExportSchema = new mongoose.Schema({
  user: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  exportType: {
    type: String,
    enum: ['buyer_orders', 'customer_orders', 'all_orders'],
    required: true
  },
  fileName: String,
  geoJSONData: Object,
  fileSize: Number,
  status: {
    type: String,
    enum: ['generating', 'ready', 'expired', 'error'],
    default: 'generating'
  },
  downloadCount: {
    type: Number,
    default: 0
  },
  lastDownloaded: Date,
  expiresAt: Date,
  filters: Object
}, {
  timestamps: true
});

geoJSONExportSchema.statics.generateForUser = async function(userId, exportType, filters = {}) {
  const Order = mongoose.model('Order');

  let query = {};

  // Query based on export type
  if (exportType === 'buyer_orders') {
    query.buyer = userId;
  } else if (exportType === 'customer_orders') {
    query.buyer = userId;
  } else if (exportType === 'all_orders') {
    query.$or = [{ buyer: userId }, { seller: userId }];
  }

  // Only export confirmed orders
  query.status = 'confirmed';

  // Apply additional filters if provided
  if (filters.dateFrom) {
    query.createdAt = { ...query.createdAt, $gte: new Date(filters.dateFrom) };
  }
  if (filters.dateTo) {
    query.createdAt = { ...query.createdAt, $lte: new Date(filters.dateTo) };
  }

  const orders = await Order.find(query)
    .populate('book', 'title author price category')
    .populate('buyer', 'profile phone')
    .populate('seller', 'profile phone')
    .sort({ createdAt: -1 })
    .lean();

  const features = orders
    .filter(o => o.buyerLocation && o.buyerLocation.coordinates && o.buyerLocation.coordinates.length === 2)
    .map(order => ({
      type: 'Feature',
      geometry: {
        type: 'Point',
        coordinates: order.buyerLocation.coordinates
      },
      properties: {
        orderId: order._id.toString(),
        bookTitle: order.book?.title || 'Unknown',
        bookAuthor: order.book?.author || 'Unknown',
        bookCategory: order.book?.category || 'Unknown',
        quantity: order.quantity,
        totalPrice: order.totalPrice,
        status: order.status,
        locationName: order.buyerLocation.name || 'Delivery Point',
        address: order.buyerLocation.address || '',
        buyerName: order.buyer?.profile ? `${order.buyer.profile.firstName || ''} ${order.buyer.profile.lastName || ''}`.trim() : 'Customer',
        buyerPhone: order.buyer?.phone || '',
        sellerName: order.seller?.profile ? `${order.seller.profile.firstName || ''} ${order.seller.profile.lastName || ''}`.trim() : 'Seller',
        sellerPhone: order.seller?.phone || '',
        orderDate: order.createdAt,
        notes: order.buyerNotes || ''
      }
    }));

  const geoJSONData = {
    type: 'FeatureCollection',
    features: features,
    properties: {
      exportType,
      generatedAt: new Date().toISOString(),
      totalFeatures: features.length
    }
  };

  const fileName = `${exportType}_${userId}_${Date.now()}.geojson`;

  const exportRecord = await this.create({
    user: userId,
    exportType,
    fileName,
    geoJSONData,
    fileSize: JSON.stringify(geoJSONData).length,
    status: 'ready',
    expiresAt: new Date(Date.now() + 30 * 24 * 60 * 60 * 1000),
    filters
  });

  console.log(`✅ Created ${exportType} export with ${features.length} features for user ${userId}`);
  return exportRecord;
};

module.exports = mongoose.model('GeoJSONExport', geoJSONExportSchema);

File: lib/models/geojson_export.dart
class GeoJSONExport {
  final String id;
  final String fileName;
  final int fileSize;
  final String status;
  final String exportType;
  final DateTime createdAt;
  final DateTime expiresAt;
  final int downloadCount;
  final DateTime? lastDownloaded;
  final int? featureCount;

  GeoJSONExport({
    required this.id,
    required this.fileName,
    required this.fileSize,
    required this.status,
    required this.exportType,
    required this.createdAt,
    required this.expiresAt,
    required this.downloadCount,
    this.lastDownloaded,
    this.featureCount,
  });

  factory GeoJSONExport.fromJson(Map<String, dynamic> json) {
    return GeoJSONExport(
      id: json['id'] ?? '',
      fileName: json['fileName'] ?? '',
      fileSize: json['fileSize'] ?? 0,
      status: json['status'] ?? 'generating',
      exportType: json['exportType'] ?? 'buyer_orders',
      createdAt: DateTime.parse(json['createdAt']),
      expiresAt: DateTime.parse(json['expiresAt']),
      downloadCount: json['downloadCount'] ?? 0,
      lastDownloaded: json['lastDownloaded'] != null ? DateTime.parse(json['lastDownloaded']) : null,
      featureCount: json['featureCount'],
    );
  }

  bool get isReady => status == 'ready';
  bool get isExpired => status == 'expired' || expiresAt.isBefore(DateTime.now());

  String get fileSizeFormatted {
    if (fileSize < 1024) return '$fileSize B';
    if (fileSize < 1024 * 1024) return '${(fileSize / 1024).toStringAsFixed(1)} KB';
    return '${(fileSize / (1024 * 1024)).toStringAsFixed(1)} MB';
  }
}

class GeoJSONFeature {
  final String type;
  final Geometry geometry;
  final Properties properties;

  GeoJSONFeature({
    required this.type,
    required this.geometry,
    required this.properties,
  });

  factory GeoJSONFeature.fromJson(Map<String, dynamic> json) {
    return GeoJSONFeature(
      type: json['type'] ?? 'Feature',
      geometry: Geometry.fromJson(json['geometry']),
      properties: Properties.fromJson(json['properties']),
    );
  }
}

class Geometry {
  final String type;
  final List<double> coordinates;

  Geometry({
    required this.type,
    required this.coordinates,
  });

  factory Geometry.fromJson(Map<String, dynamic> json) {
    return Geometry(
      type: json['type'] ?? 'Point',
      coordinates: List<double>.from(json['coordinates'].map((x) => x.toDouble())),
    );
  }

  double get latitude => coordinates[1];
  double get longitude => coordinates[0];
}

class Properties {
  final String? orderId;
  final String? bookTitle;
  final String? bookAuthor;
  final int? quantity;
  final double? totalPrice;
  final String? orderType;
  final String? status;
  final String? locationName;
  final String? address;
  final String? contactName;
  final String? contactPhone;
  final String? pointType;
  final DateTime? orderDate;
  final String? notes;

  Properties({
    this.orderId,
    this.bookTitle,
    this.bookAuthor,
    this.quantity,
    this.totalPrice,
    this.orderType,
    this.status,
    this.locationName,
    this.address,
    this.contactName,
    this.contactPhone,
    this.pointType,
    this.orderDate,
    this.notes,
  });

  factory Properties.fromJson(Map<String, dynamic> json) {
    return Properties(
      orderId: json['orderId'],
      bookTitle: json['bookTitle'],
      bookAuthor: json['bookAuthor'],
      quantity: json['quantity'],
      totalPrice: json['totalPrice']?.toDouble(),
      orderType: json['orderType'],
      status: json['status'],
      locationName: json['locationName'],
      address: json['address'],
      contactName: json['contactName'],
      contactPhone: json['contactPhone'],
      pointType: json['pointType'],
      orderDate: json['orderDate'] != null ? DateTime.parse(json['orderDate']) : null,
      notes: json['notes'],
    );
  }
}

class GeoJSONData {
  final String type;
  final List<GeoJSONFeature> features;

  GeoJSONData({
    required this.type,
    required this.features,
  });

  factory GeoJSONData.fromJson(Map<String, dynamic> json) {
    return GeoJSONData(
      type: json['type'] ?? 'FeatureCollection',
      features: (json['features'] as List).map((feature) => GeoJSONFeature.fromJson(feature)).toList(),
    );
  }

  int get featureCount => features.length;
}
